test/jdk/sun/net/www/protocol/http/AuthCache.java

```diff
@@ -0,0 +1,184 @@
+         * @return
+         */
+        public synchronized boolean wasCalled() {
+            if (errorMsg != null)
+                throw new RuntimeException(errorMsg);
+
+            return wasCalled.getAndSet(false);
+        }
+        protected synchronized PasswordAuthentication getPasswordAuthentication() {
+            if (!getRequestingPrompt().equals(realm)) {
+                errorMsg = String.format("Error: %s expected as realm, received %s", realm, getRequestingPrompt());
+            }
+            wasCalled.set(true);
+            return new PasswordAuthentication(username, password.toCharArray());
+        }
+    }
+
+    static final HttpHandler handler = (HttpExchange exch) -> {
+        exch.sendResponseHeaders(200, -1);
+        exch.close();
+    };
+
+    static class ServerAuth extends BasicAuthenticator {
+        private final String user, pass;
+
+        ServerAuth(String realm, String user, String pass) {
+            super(realm);
+            this.user = user;
+            this.pass = pass;
+        }
+
+        @Override
+        public boolean checkCredentials(String username, String password) {
+            return username.equals(user) && password.equals(pass);
+        }
+    }
+    /**
+     * Creates two Authenticators and two realms ("r1" and "r2")
+     * "r1" uses context "/path1" credentials = user1/pass1
+     * "r2" uses context "/path2" credentials = user2/pass2
+     *
+     * 1) Send request to "r1" and "r2" expect both authenticators to be called
+     *    cache size should be 4
+     *
+     * 2) Send request to "r1" and "r2". Authenticators should not be called (cache)
+     *
+     * 3) Clear reference to "r1" and call gc.
+     *    cache size should be 2
+     *
+     * 4) Send request to "r1" and "r2". "r1" auth should be called, but not "r2"
+     *    cache size should be 4
+     */
+    public static void main(String[] args) throws IOException {
+        var clauth1 = new ClientAuth("r1", "user1", "pass1");
+        var clauth2 = new ClientAuth("r2", "user2", "pass2");
+        var server = HttpServer.create(new InetSocketAddress(InetAddress.getLoopbackAddress(), 0), 0);
+        var ctx1 = server.createContext("/path1", handler);
+        ctx1.setAuthenticator(new ServerAuth("r1", "user1", "pass1"));
+
+        var ctx2 = server.createContext("/path2", handler);
+        ctx2.setAuthenticator(new ServerAuth("r2", "user2", "pass2"));
+        var addr = server.getAddress();
+        var url1 = URI.create("http://" + addr.getHostName() + ":" + addr.getPort() + "/path1/").toURL();
+        var url2 = URI.create("http://" + addr.getHostName() + ":" + addr.getPort() + "/path2/").toURL();
+        server.start();
+
+        sendRequest(url1, url2, clauth1, clauth2, true, true);
+        checkCacheSize(4);
+        sendRequest(url1, url2, clauth1, clauth2, false, false);
+        checkCacheSize(4);
+        clauth1 = null;
+        System.gc();
+        delay(1);
+        System.gc();
```

Comment:
I suggest creating a weak reference (clauth1) with a reference queue, and then pull the reference queue in a loop (with a smaller sleep timeout), and calling System.gc() each time the timeout elapses.

```
        ReferenceQueue<Authenticator> queue = new ReferenceQueue();
        WeakReference<Authenticator> ref = new WeakReference<>(clauth1, queue);
        clauth1 = null;
        Reference<Authenticator> ref2 = null;
        while ((ref2 = queue.remove(100)) != ref) {
            System.gc();
        }
```

Alternatively you could try to use something like:

```
        WeakReference<Authenticator> ref = new WeakReference<>(clauth1);
        clauth1 = null;
        ForceGC.await(() -> ref.refersTo(null))
```

Reply:
Good idea. The second approach seems quite tidy.